feat: Improve `rg.log` function

frascuchon opened the pull request:
# Description

Allow log data batches concurrently

- Accept `num_threads` to log batches concurrently    
- Add retries when an `httpx.TransportError` occurs


Close partially #2533

**Type of change**

(Please delete options that are not relevant. Remember to title the PR according to the type of change)

- [x] Improvement (change adding some improvement to an existing functionality)

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes. And ideally, reference `tests`)

TDB

**Checklist**

- [x] I have merged the original branch into my forked branch
- [ ] I added relevant documentation
- [x] follows the style guidelines of this project
- [x] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

frascuchon committed:
Allow log data batches concurrently

- Accept `num_threads` to log batches concurrently
- All batches will be processed even if errores were found
- For those with errors a more descriptive error will be raised

frascuchon committed:
Adapt tests

frascuchon committed:
Upgrade changelog

frascuchon committed:
Fixing tests

frascuchon committed:
Other log improvements

- Increase default timeout
- retry when a `httpx.TransportError` error occurs

frascuchon committed:
Update CHANGELOG

frascuchon committed:
Set num_threads=0 for non threading at all

frascuchon committed:
Add basic test

frascuchon committed:
Remove extra files

frascuchon commented on src/argilla/client/client.py:
```diff
@@ -391,27 +349,41 @@ async def log_async(
                 progress_bar.update(task, advance=len(batch))
+                return bulk_result.processed, bulk_result.failed
+
+            if num_threads > 1:
```
```suggestion
            if num_threads >= 1:
```

frascuchon committed:
Update src/argilla/client/client.py

frascuchon committed:
Merge branch 'develop' into feat/improve-rg.log-functions

davidberenstein1957 commented on src/argilla/client/api.py:
```diff
@@ -112,10 +112,11 @@ def log(
     verbose: bool = True,
     background: bool = False,
     chunk_size: Optional[int] = None,
+    num_threads: int = 0,
```
Hi, thanks for fixing this. I would also like to see `max_retries` and the `backoff` variables here.

frascuchon replied:
Agree with `max_retries`, I forget to add it. For the backoff not sure what you want to expose really? I would like to start with an MVP allowing the setup of at least the retries before stopping. We can consider exposing other parameters later. wdyt @davidberenstein1957 ?

tomaarsen replied:
`backoff` likely involves e.g. "expo", "constant", "fibo", etc. from https://github.com/litl/backoff
I think this is very overkill. Few people would ever change this from the default. I'd stick with `max_retries` as that one's easy to 1) use, 2) understand and 3) implement.

davidberenstein1957 replied:
IMO, we could add max_retries in combination with a default backoff 10, 20, 30

davidberenstein1957 commented on src/argilla/client/client.py:
```diff
@@ -239,10 +220,7 @@ def copy(self, dataset: str, name_of_copy: str, workspace: str = None):
+            json_body=CopyDatasetRequest(name=name_of_copy, target_workspace=workspace),
         )
 
     def delete(self, name: str, workspace: Optional[str] = None):
```
I would also like to see the `backoff` variables here.

frascuchon replied:
We should apply changes step by step. We can consider adding a backoff mechanism to another method in a separate PR. Otherwise, a lot of changes will be included in the same PR, which can be a great bug farm. :-)

davidberenstein1957 replied:
https://github.com/argilla-io/argilla/issues/2687

davidberenstein1957 replied:
I added it to `rg.load` which seam the most relevant to me.

frascuchon replied:
For `rg.load`, things could be a bit different. For instance, we should decrease the batch size, or we should prefetch some data before splitting and parallelizing the data loading. But yes. we can have a similar approach to improve also that method

tomaarsen replied:
Decreasing batch size on failure seems very smart for `rg.load` in particular.

davidberenstein1957 commented on src/argilla/client/client.py:
```diff
@@ -423,6 +395,53 @@ async def log_async(
+        batch_size: int = 100,
+        verbose: bool = True,
+        chunk_size: Optional[int] = None,
+        num_threads: int = 1,
```
I would also like to see `max_retries` and the `backoff` variables here.

davidberenstein1957 reviewed (commented):
Hi @frascuchon looks good. I would like to see the `max_retries`, and `backoff` in the other log functions, but ideally, they should also be added to the `rg.load`.

frascuchon committed:
Remove log_async method from client class

frascuchon committed:
Using log(background=True) for log_async with deprecation warning

frascuchon committed:
Merge branch 'develop' into feat/improve-rg.log-functions

tomaarsen commented on CHANGELOG.md:
```diff
@@ -7,9 +7,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+## Changed
 
 - Argilla quickstart image dependencies are externalized into `quickstart.requirements.txt`. See [#2666](https://github.com/argilla-io/argilla/pull/2666)
+- The `rg.log_async` is deprecated and will be removed in next minor release.
```
Doesn't this fall under a "Deprecated" section?
![image](https://user-images.githubusercontent.com/37621491/231721755-665e30d7-43b0-4350-9793-db3d792232a1.png)

frascuchon replied:
Totally

tomaarsen commented on CHANGELOG.md:
```diff
@@ -7,9 +7,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Argilla quickstart image dependencies are externalized into `quickstart.requirements.txt`. See [#2666](https://github.com/argilla-io/argilla/pull/2666)
+- The `rg.log_async` is deprecated and will be removed in next minor release.
+- The `rg.log` will compute all batches and raise an error for all failed batches.
 
```
Perhaps mention that the batch size is now 100 under "Changed"? May be too small of a change to mention, though.

tomaarsen commented on src/argilla/client/client.py:
```diff
@@ -263,10 +241,12 @@ def log(
         background: bool = False,
         chunk_size: Optional[int] = None,
+        num_threads: int = 0,
+        max_retries: int = 3,
```
```suggestion
        num_threads: int = 0,
        max_retries: int = 3,
        chunk_size: Optional[int] = None,
```
Let's keep deprecated parameters at the bottom.

tomaarsen commented on src/argilla/client/client.py:
```diff
@@ -283,59 +263,38 @@ def log(
+            num_threads: If > 0, will use num_thread separate number threads to batches, sending data concurrently.
+                Default to `0`, which means no threading at all.
+            max_retries: Number of retries when logging a batch of records if a `httpx.TransportError` occurs.
+                Default `3`
```
```suggestion
            num_threads: If > 0, will use num_thread separate number threads to batches, sending data concurrently.
                Default to `0`, which means no threading at all.
            max_retries: Number of retries when logging a batch of records if a `httpx.TransportError` occurs.
                Default `3`
            chunk_size: DEPRECATED! Use `batch_size` instead.
```
Moving the deprecated parameter to the bottom.

tomaarsen commented on src/argilla/client/client.py:
```diff
@@ -283,59 +263,38 @@ def log(
-            Summary of the response from the REST API
+        if workspace is not None:
+            self.set_workspace(workspace)
 
```
```suggestion
```
We already have this a few lines below:
https://github.com/argilla-io/argilla/blob/a23f9321cfa819f1e3a5fab7dd72a1047c5fbd15/src/argilla/client/client.py#L301-L302

frascuchon replied:
Done

tomaarsen commented on src/argilla/client/client.py:
```diff
@@ -389,27 +348,41 @@ async def log_async(
+            processed, failed = 0, 0
+            for processed_batch, failed_batch in results:
+                processed += processed_batch
+                failed += failed_batch
```
I feel like this could be written more concisely, but it ends up being a mess of `zip`'s and `sum`'s 😅 
So we can just keep it like this.

frascuchon replied:
Yes, I've refactored a bit this code block. Thanks for remarking it

tomaarsen reviewed (changes requested):
Some small nitpicks. Looks good otherwise, I'm always glad to see a PR that removes more code than it adds.

frascuchon committed:
Apply CHANGELOG suggestions

- Use the Deprecated section for CHANGELOG
- Include new default batch size for rg.log

frascuchon committed:
Remove duplicated statement

frascuchon commented on CHANGELOG.md:
```diff
@@ -7,9 +7,19 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - Argilla quickstart image dependencies are externalized into `quickstart.requirements.txt`. See [#2666](https://github.com/argilla-io/argilla/pull/2666)
+- The `rg.log` computes all batches and raise an error for all failed batches.
+- The `rg.log` default batch size is now 100.
```
```suggestion
- The default batch size for `rg.log` is now 100.
```

frascuchon committed:
Apply suggestions from code review

Co-authored-by: Tom Aarsen <37621491+tomaarsen@users.noreply.github.com>

frascuchon committed:
Merge remote-tracking branch 'origin/develop' into feat/improve-rg.log-functions

frascuchon committed:
Merge branch 'feat/improve-rg.log-functions' of github.com:recognai/rubrix into feat/improve-rg.log-functions

frascuchon committed:
Merge branch 'develop' into feat/improve-rg.log-functions

frascuchon commented on CHANGELOG.md:
```diff
@@ -7,14 +7,22 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [Unreleased]
 
-###  Added
+## Added
```
```suggestion
### Added
```

frascuchon commented on CHANGELOG.md:
```diff
@@ -7,14 +7,22 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+- The `rg.log` computes all batches and raise an error for all failed batches.
+- The default batch size for `rg.log` is now 100.
+
+## Deprecated
```
```suggestion
### Deprecated
```

frascuchon committed:
Apply suggestions from code review

tomaarsen reviewed (approved):
Looks great, I think this is all set
